from PySide6.QtWidgets import (
    QTextBrowser, QPushButton, QFileDialog, QComboBox, QLabel, QWidget
)
from PySide6.QtCore import QTimer
from PySide6.QtGui import QColor, QPalette

# å¯¼å…¥ ui_utils ä¸­çš„ BaseTabWidget, WorkerThread (éœ€è¦ç¡®ä¿è¿™äº›ç±»æ˜¯å­˜åœ¨çš„)
from ui_utils import BaseTabWidget, WorkerThread


class StatsTabWidget(BaseTabWidget):
    # ğŸŒŸ æ–°å¢ï¼šç”¨äº HTML æ¸²æŸ“çš„è‡ªé€‚åº”é¢œè‰²å±æ€§ ğŸŒŸ
    EYE_CARE_COLOR = "#FFFFF0"
    TEXT_COLOR = "black"

    def __init__(self, qingshilu_service, parent=None):
        super().__init__("stats_tab.ui", parent)
        self.service = qingshilu_service

        # ğŸŒŸ å…³é”®ï¼šè°ƒç”¨ä¸»é¢˜é€‚åº”é€»è¾‘ ğŸŒŸ
        self.apply_theme_adaptation()

        # å‡è®¾ QingShiluService å°† DataModel å®ä¾‹å­˜å‚¨åœ¨åä¸º 'model' çš„å±æ€§ä¸­ã€‚
        self.category_structure = self.service.model.get_category_structure()

        # å½“å‰é€‰æ‹©çš„åˆ†ç±»é”®
        self.current_l1_key = None
        self.current_l2_name = None
        self.current_l3_name = None

        # å­˜å‚¨ worker çº¿ç¨‹å®ä¾‹ï¼Œé˜²æ­¢å…¶è¢«é”€æ¯å¯¼è‡´ç¨‹åºå´©æºƒ (SIGABRT)
        self.worker = None

        self.notificationLabel = None

        self._find_widgets()
        self._setup_ui_fixes()
        self.connect_signals()
        self._init_comboboxes()

        QTimer.singleShot(100, self.load_stats_data)

    # -----------------------------------------------------------
    # ğŸŒŸ æ–°å¢ï¼šä¸»é¢˜é€‚åº”é€»è¾‘ ğŸŒŸ
    # -----------------------------------------------------------
    def apply_theme_adaptation(self):
        """
        æ ¹æ®ç³»ç»Ÿä¸»é¢˜ï¼Œè®¾ç½®æµ…è‰²æ¨¡å¼ä¸‹çš„èƒŒæ™¯è‰²ä¸ºç±³ç™½è‰² (#FFFFF0)ï¼Œ
        æ·±è‰²æ¨¡å¼ä¸‹ä½¿ç”¨ Qt çš„ç³»ç»Ÿé»˜è®¤èƒŒæ™¯è‰²ï¼Œå®ç°è‡ªé€‚åº”ã€‚
        """
        palette = self.palette()

        # è·å–ç³»ç»Ÿçª—å£é¢œè‰²ä½œä¸ºåˆ¤æ–­ä¾æ®
        window_color = palette.color(QPalette.Window)

        # åˆ¤æ–­æ˜¯å¦ä¸ºæ·±è‰²æ¨¡å¼ï¼šåŸºäºäº®åº¦çš„å¯å‘å¼åˆ¤æ–­
        is_dark_mode = window_color.red() < 128 and window_color.green() < 128 and window_color.blue() < 128

        if not is_dark_mode:
            # â˜€ï¸ æµ…è‰²æ¨¡å¼ï¼šå¼ºåˆ¶ä½¿ç”¨ç±³ç™½è‰² (#FFFFF0)
            custom_color = QColor("#FFFFF0")

            # è®¾ç½®æ•´ä¸ª Tab é¡¶å±‚ Widget çš„èƒŒæ™¯
            palette.setColor(QPalette.Window, custom_color)
            self.setAutoFillBackground(True)
            self.setPalette(palette)

            self.EYE_CARE_COLOR = "#FFFFF0"
            self.TEXT_COLOR = "black"
        else:
            # ğŸŒ™ æ·±è‰²æ¨¡å¼ï¼šä¿æŒé»˜è®¤ï¼Œè·Ÿéšç³»ç»Ÿä¸»é¢˜
            self.setAutoFillBackground(False)

            # åœ¨æ·±è‰²æ¨¡å¼ä¸‹ï¼Œä½¿ç”¨ç³»ç»ŸèƒŒæ™¯è‰²ä½œä¸ºæ–‡æœ¬æµè§ˆå™¨çš„èƒŒæ™¯
            dark_mode_bg_color = palette.color(QPalette.Base).name()  # ä½¿ç”¨ Base è€Œé Window é¢œè‰²ä½œä¸ºå†…å®¹åŒºåŸŸèƒŒæ™¯
            self.EYE_CARE_COLOR = dark_mode_bg_color
            self.TEXT_COLOR = "white"  # æ·±è‰²æ¨¡å¼ä¸‹æ–‡æœ¬é¢œè‰²ä¸ºç™½è‰²

    def _find_widgets(self):
        """æŸ¥æ‰¾ UI æ§ä»¶"""
        self.statsTextBrowser = self.findChild(QTextBrowser, "statsTextBrowser")
        self.exportCsvButton = self.findChild(QPushButton, "exportCsvButton")
        self.refreshButton = self.findChild(QPushButton, "refreshButton")
        self.level1ComboBox = self.findChild(QComboBox, "level1ComboBox")
        self.level2ComboBox = self.findChild(QComboBox, "level2ComboBox")
        self.level3ComboBox = self.findChild(QComboBox, "level3ComboBox")
        self.statsTitleLabel = self.findChild(QLabel, "statsTitleLabel")  # ç¡®ä¿æ ‡é¢˜æ ‡ç­¾è¢«æ‰¾åˆ°

        # æŸ¥æ‰¾é€šçŸ¥ QLabel
        self.notificationLabel = self.findChild(QLabel, "notificationLabel")
        if self.notificationLabel:
            self.notificationLabel.hide()  # é»˜è®¤éšè—

        # ğŸŒŸ å…³é”®ï¼šå¯¹ statsTextBrowser åº”ç”¨è‡ªé€‚åº”æ ·å¼ ğŸŒŸ
        if self.statsTextBrowser:
            # è®¾ç½® statsTextBrowser çš„èƒŒæ™¯å’Œæ–‡å­—é¢œè‰²
            self.statsTextBrowser.setStyleSheet(
                f"""
                QTextBrowser {{
                    background-color: {self.EYE_CARE_COLOR}; 
                    color: {self.TEXT_COLOR};
                    border: 1px solid #ccc; 
                    padding: 10px;
                }}
                """
            )

        # ğŸŒŸ å¯¹ statsTitleLabel é‡æ–°è®¾ç½® HTML é¢œè‰²ä»¥ç¡®ä¿æ·±è‰²æ¨¡å¼ä¸‹å¯è§ ğŸŒŸ
        if self.statsTitleLabel and self.TEXT_COLOR == "white":
            self.statsTitleLabel.setText(
                '<h2 style="color:white;">å·²åˆ†ç±»æ¡æ–‡ç»Ÿè®¡ä¸å¯¼å‡º</h2>'
            )

    def _setup_ui_fixes(self):
        """ä¿®å¤ï¼šè®¾ç½® ComboBox çš„å¤§å°è°ƒæ•´ç­–ç•¥å’Œæœ€å°å®½åº¦ä»¥ç¡®ä¿é•¿æ–‡æœ¬æ˜¾ç¤ºå®Œæ•´"""
        min_width = 180  # å¢å¤§æœ€å°å®½åº¦ï¼Œæ›´å¥½åœ°é€‚åº”ä¸­æ–‡é•¿æ–‡æœ¬

        if self.level1ComboBox:
            self.level1ComboBox.setMinimumWidth(min_width)
            # å…è®¸æ ¹æ®å†…å®¹è‡ªåŠ¨è°ƒæ•´å®½åº¦
            self.level1ComboBox.setSizeAdjustPolicy(QComboBox.AdjustToContents)

        if self.level2ComboBox:
            self.level2ComboBox.setMinimumWidth(min_width)
            self.level2ComboBox.setSizeAdjustPolicy(QComboBox.AdjustToContents)

        if self.level3ComboBox:
            self.level3ComboBox.setMinimumWidth(min_width)
            self.level3ComboBox.setSizeAdjustPolicy(QComboBox.AdjustToContents)

    def connect_signals(self):
        """è¿æ¥ä¿¡å·æ§½"""
        if self.exportCsvButton:
            self.exportCsvButton.clicked.connect(self._handle_export_csv)
        if self.refreshButton:
            self.refreshButton.clicked.connect(self.load_stats_data)

        # è”åŠ¨ä¿¡å·ï¼Œä½¿ç”¨ try-except å¿½ç•¥æœªè¿æ¥çš„è­¦å‘Š
        if self.level1ComboBox:
            try:
                self.level1ComboBox.currentIndexChanged.disconnect(self._update_level2_combobox)
            except Exception:
                pass
            self.level1ComboBox.currentIndexChanged.connect(self._update_level2_combobox)

        if self.level2ComboBox:
            try:
                self.level2ComboBox.currentIndexChanged.disconnect(self._update_level3_combobox)
            except Exception:
                pass
            self.level2ComboBox.currentIndexChanged.connect(self._update_level3_combobox)

    def _init_comboboxes(self):
        """åˆå§‹åŒ–ä¸€çº§åˆ†ç±»ä¸‹æ‹‰æ¡†"""
        self.level1ComboBox.blockSignals(True)
        self.level1ComboBox.clear()

        # é»˜è®¤é¡¹ï¼šå…¨éƒ¨
        self.level1ComboBox.addItem("å…¨éƒ¨", None)

        # ä»åˆ†ç±»ç»“æ„ä¸­æ·»åŠ ä¸€çº§åˆ†ç±»
        for l1_key, l1_data in self.category_structure.items():
            l1_name = l1_data.get('name', str(l1_key))
            self.level1ComboBox.addItem(l1_name, l1_key)

        self.level1ComboBox.setCurrentIndex(0)
        self.level1ComboBox.blockSignals(False)

        self._update_level2_combobox()

    def _update_level2_combobox(self):
        """æ ¹æ®ä¸€çº§åˆ†ç±»çš„é€‰æ‹©æ›´æ–°äºŒçº§åˆ†ç±»ä¸‹æ‹‰æ¡†"""
        self.level2ComboBox.blockSignals(True)
        self.level2ComboBox.clear()

        self.current_l1_key = self.level1ComboBox.currentData()

        self.level2ComboBox.addItem("å…¨éƒ¨", None)

        if self.current_l1_key is not None:
            l1_structure = self.category_structure.get(self.current_l1_key)
            if l1_structure:
                for l2_name in l1_structure.get('levels', {}).keys():
                    self.level2ComboBox.addItem(l2_name, l2_name)

        self.level2ComboBox.setCurrentIndex(0)
        self.level2ComboBox.blockSignals(False)
        self._update_level3_combobox()

    def _update_level3_combobox(self):
        """æ ¹æ®äºŒçº§åˆ†ç±»çš„é€‰æ‹©æ›´æ–°ä¸‰çº§åˆ†ç±»ä¸‹æ‹‰æ¡†"""
        self.level3ComboBox.blockSignals(True)
        self.level3ComboBox.clear()

        self.current_l2_name = self.level2ComboBox.currentData()

        self.level3ComboBox.addItem("å…¨éƒ¨", None)

        if self.current_l1_key is not None and self.current_l2_name is not None:
            l1_structure = self.category_structure.get(self.current_l1_key)
            if l1_structure:
                l2_structure = l1_structure.get('levels', {}).get(self.current_l2_name)

                if l2_structure:
                    for l3_name in l2_structure.keys():
                        self.level3ComboBox.addItem(l3_name, l3_name)

        self.level3ComboBox.setCurrentIndex(0)
        self.level3ComboBox.blockSignals(False)

        self.current_l3_name = self.level3ComboBox.currentData()
        self.load_stats_data()

    def _get_filter_key(self):
        """æ ¹æ®ä¸‹æ‹‰æ¡†é€‰æ‹©æ„é€  Service è¦æ±‚çš„ç­›é€‰é”®"""
        l1_key = self.level1ComboBox.currentData()
        l2_name = self.level2ComboBox.currentData()
        l3_name = self.level3ComboBox.currentData()

        if l1_key is None:
            return None

        if l2_name is None:
            return str(l1_key)

        if l3_name is None:
            return f"{l1_key}/{l2_name}"

        return f"{l1_key}/{l2_name}/{l3_name}"

    def load_stats_data(self):
        """ä» Service è·å–åˆ†ç±»æ•°æ®å¹¶æ¸²æŸ“ç»Ÿè®¡ä¿¡æ¯ (æ ¹æ®ç­›é€‰)"""
        filter_key = self._get_filter_key()

        try:
            stats_data = self.service.model.get_classified_stats(filter_key)
        except Exception as e:
            self.statsTextBrowser.setHtml(
                f"<h3 style='color:red;'>æ•°æ®åŠ è½½é”™è¯¯ï¼š</h3><p>è¯·æ£€æŸ¥ service.model.get_classified_stats æ–¹æ³•å’Œæ•°æ®ç»“æ„ã€‚é”™è¯¯ä¿¡æ¯ï¼š{e}</p>")
            return

        # ğŸŒŸ å…³é”®ï¼šä½¿ç”¨è‡ªé€‚åº”æ–‡æœ¬é¢œè‰² ğŸŒŸ
        text_color_style = f"color:{self.TEXT_COLOR};"

        # è°ƒæ•´ HTML æ¨¡æ¿ä»¥ä½¿ç”¨è‡ªé€‚åº”é¢œè‰²
        # æ ‡é¢˜é¢œè‰²ä½¿ç”¨ #2d5a3e ä¿æŒè§†è§‰ä¸€è‡´æ€§ï¼Œä½†åŸºç¡€æ–‡æœ¬é¢œè‰²ä½¿ç”¨è‡ªé€‚åº”çš„ self.TEXT_COLOR
        html = f'<h3 style="{text_color_style} color:#2d5a3e;">æ€»è®¡å·²åˆ†ç±»æ¡æ–‡: <span style="font-size: 1.2em;" id="total_count">0</span> æ¡</h3><hr style="border-color:{self.TEXT_COLOR};">'
        total_count = 0
        detail_html = f'<h4 style="margin-top: 15px; {text_color_style} color:#555;">å½“å‰ç­›é€‰ç»“æœå±‚çº§ç»Ÿè®¡</h4>'

        if stats_data:
            for l1_key, l1_data in stats_data.items():
                l1_name = l1_data.get('name', f"L1 Key {l1_key}")
                l1_total = l1_data['count']
                total_count += l1_total

                # L1 æ ‡é¢˜
                detail_html += f'<div style="margin-top: 15px; {text_color_style}">'
                detail_html += f'<h4><b style="color:#1e40af;">[{l1_name}] ({l1_total} æ¡)</b></h4>'

                if 'levels' in l1_data:
                    detail_html += '<ul style="list-style: disc; margin-left: 20px;">'
                    for l2_name, l2_data in l1_data['levels'].items():
                        l2_count = l2_data['count']
                        # L2 åˆ—è¡¨é¡¹
                        detail_html += f'<li><b>{l2_name} ({l2_count} æ¡)</b>'

                        # L3 æ¸²æŸ“
                        if 'levels' in l2_data and l2_data['levels']:
                            # L3 æ ‡ç­¾çš„èƒŒæ™¯è‰²ä¿æŒæµ…è‰²ï¼ˆ#e0f2feï¼‰ä»¥å¢åŠ å¯¹æ¯”åº¦ï¼Œå­—ä½“ç”¨é»‘è‰²
                            l3_bg_color = "#e0f2fe"
                            l3_text_color = "black"

                            detail_html += '<div style="margin-left: 20px; margin-top: 5px; line-height: 2.0;">'
                            l3_spans = []
                            for l3_name, l3_count in l2_data['levels'].items():
                                l3_spans.append(
                                    f'<span style="padding: 2px 5px; background-color: {l3_bg_color}; color:{l3_text_color}; border-radius: 4px; margin-right: 10px; white-space: nowrap;">{l3_name} ({l3_count} æ¡)</span>')

                            # ä½¿ç”¨ç©ºæ ¼è¿æ¥ spanï¼Œè®©æµè§ˆå™¨è‡ªåŠ¨æ¢è¡Œ
                            detail_html += ' '.join(l3_spans)
                            detail_html += '</div>'

                        detail_html += "</li>"
                    detail_html += "</ul>"
                detail_html += "</div>"

        # æ›´æ–°æ€»è®¡è®¡æ•°ï¼Œç„¶åæ‹¼æ¥è¯¦ç»†å†…å®¹
        final_html = html.replace('<span style="font-size: 1.2em;" id="total_count">0</span>',
                                  f'<span style="font-size: 1.2em; color: #2d5a3e;">{total_count}</span>')
        final_html += detail_html

        if self.statsTextBrowser:
            self.statsTextBrowser.setHtml(final_html)

    # -----------------------------------------------------------
    # show_notification æ–¹æ³•
    # -----------------------------------------------------------
    def show_notification(self, message: str, type: str = 'info'):
        """æ˜¾ç¤ºä¸´æ—¶çš„åº•éƒ¨é€šçŸ¥ï¼Œ3ç§’åè‡ªåŠ¨éšè— (å–ä»£ CustomToast)"""
        if not self.notificationLabel:
            return

        # æ ·å¼å‚è€ƒ BatchTabWidgetï¼Œå¹¶ç•¥ä½œç¾åŒ–è°ƒæ•´
        style = "padding: 8px 15px; border-radius: 4px; font-weight: bold; text-align: center; margin-top: 5px;"

        # å®šä¹‰é¢œè‰²ä¸»é¢˜
        if type == 'error':
            # çº¢è‰²ä¸»é¢˜
            style += "background-color: #f8d7da; color: #721c24; border: 1px solid #f5c6cb;"
        elif type == 'warning':
            # é»„è‰²ä¸»é¢˜
            style += "background-color: #fff3cd; color: #856404; border: 1px solid #ffeeba;"
        elif type == 'success':
            # æˆåŠŸä¸»é¢˜ (ç»¿è‰²)
            style += "background-color: #d4edda; color: #155724; border: 1px solid #c3e6cb;"
        elif type == 'info':
            # ä¿¡æ¯ä¸»é¢˜ (è“è‰²/ç°è‰²)
            style += "background-color: #e0f7fa; color: #004d40; border: 1px solid #b2ebf2;"
        else:
            # é»˜è®¤ä¸»é¢˜
            style += "background-color: #e2e3e5; color: #383d41; border: 1px solid #d6d8db;"

        self.notificationLabel.setText(message)
        self.notificationLabel.setStyleSheet(style)
        self.notificationLabel.show()

        # 3ç§’åè‡ªåŠ¨éšè—
        QTimer.singleShot(3000, self.notificationLabel.hide)

    # -----------------------------------------------------------
    # _handle_export_csv (ä½¿ç”¨ show_notification)
    # -----------------------------------------------------------
    def _handle_export_csv(self):
        """å¤„ç†å¯¼å‡º CSV æŒ‰é’®ç‚¹å‡»ï¼Œå¯¼å‡ºå½“å‰ç­›é€‰çš„ç»“æœ"""
        filter_key = self._get_filter_key()

        # æ ¹æ®ç­›é€‰é”®ç”Ÿæˆé»˜è®¤æ–‡ä»¶å
        if filter_key is None:
            default_name = "classified_results_all.csv"
        else:
            default_name = f"classified_results_{filter_key.replace('/', '_')}.csv"

        file_path, _ = QFileDialog.getSaveFileName(
            self, "å¯¼å‡ºåˆ†ç±»ç»“æœ", default_name, "CSV æ–‡ä»¶ (*.csv);;æ‰€æœ‰æ–‡ä»¶ (*)"
        )

        if file_path:
            # æ›¿æ¢ CustomToast ä¸º show_notification (æç¤ºå¼€å§‹)
            self.show_notification("CSV æ–‡ä»¶å¯¼å‡ºå·²åœ¨åå°å¼€å§‹ï¼Œè¯·ç¨å€™...", type='info')

            # å°† worker èµ‹å€¼ç»™ self.workerï¼Œé˜²æ­¢å…¶è¢«é”€æ¯
            self.worker = WorkerThread(
                self.service.model.export_classified_data_to_csv,
                file_path=file_path,
                filter_key=filter_key
            )
            # ä½¿ç”¨ lambda æ•è·ç»“æœï¼Œå¹¶è°ƒç”¨è¾…åŠ©å‡½æ•°æ˜¾ç¤ºç»“æœ
            self.worker.result_signal.connect(lambda success: self._show_export_result(success, file_path))
            # é”™è¯¯æ—¶ä½¿ç”¨ show_notification
            self.worker.error_signal.connect(lambda err: self.show_notification(
                f"å¯¼å‡ºå¤±è´¥ï¼š{err}", type='error'
            ))

            # è¿æ¥ finished ä¿¡å·åˆ° deleteLaterï¼Œç¡®ä¿çº¿ç¨‹å®‰å…¨é€€å‡ºå’Œæ¸…ç†
            self.worker.finished.connect(self.worker.deleteLater)

            self.worker.start()

    # -----------------------------------------------------------
    # _show_export_result (ä½¿ç”¨ show_notification)
    # -----------------------------------------------------------
    def _show_export_result(self, success, file_path):
        """æ˜¾ç¤ºå¯¼å‡ºç»“æœ (ä½¿ç”¨ show_notification)"""
        if success:
            # ä½¿ç”¨ show_notification
            self.show_notification(
                f"åˆ†ç±»ç»“æœå·²æˆåŠŸå¯¼å‡ºåˆ°: {file_path}",
                type='success'
            )
        else:
            # ä½¿ç”¨ show_notification
            self.show_notification(
                "æ²¡æœ‰æ•°æ®å¯ä¾›å¯¼å‡ºï¼Œæˆ–å¯¼å‡ºå¤±è´¥ã€‚",
                type='warning'
            )
